packages/langchain-graph-retriever/src/langchain_graph_retriever/adapters/astra.py

```diff
@@ -37,61 +37,62 @@
-            return {"$and": [encoded_filter, self.encoded_user_filter]}
+    for edge in edges:
+        if isinstance(edge, MetadataEdge):
+            metadata.setdefault(edge.incoming_field, set()).add(edge.value)
+        elif isinstance(edge, IdEdge):
+            ids.add(edge.id)
         else:
-            return encoded_filter
+            raise ValueError(f"Unsupported edge {edge}")
 
-    def create_ids_query(self, ids: list[str]) -> dict[str, Any]:
-        if len(ids) == 0:
-            raise ValueError("IDs should not be empty")
-        elif len(ids) == 1:
-            # We need to avoid re-encoding in this case, since the codecs
-            # don't recognize `_id` and get confused (creating `metadata._id`).
-            return self._with_user_filter({"_id": ids[0]})
-        elif len(ids) < 100:
-            return self._with_user_filter({"_id": {"$in": ids}})
-        else:
-            raise ValueError("IDs should be less than 100, was {len(ids)}")
+    return (cast(dict[str, Iterable[Any]], metadata), ids)
 
-    def create_metadata_query(
-        self, metadata: dict[str, Iterable[Any]]
-    ) -> dict[str, Any] | None:
-        metadata = {k: v for k, v in metadata.items() if v}
-        if not metadata:
-            return None
 
-        parts = []
-        for k, v in metadata.items():
-            # If there are more than 100 values, we can't create a single `$in` query.
-            # But, we can do it for each batch of 100.
-            for v_batch in batched(v, 100):
-                batch = list(v_batch)
-                if len(batch) == 1:
-                    parts.append({k: batch[0]})
-                else:
-                    parts.append({k: {"$in": batch}})
-
-        if len(parts) == 1:
-            return self.encode_filter(parts[0])
+def _queries(
+    codec: _AstraDBVectorStoreDocumentCodec,
+    user_filters: dict[str, Any] | None,
+    *,
+    metadata: dict[str, Iterable[Any]] = {},
+    ids: Iterable[str] = (),
+) -> Iterator[dict[str, Any]]:
+    encoded_user_filters = codec.encode_filter(user_filters) if user_filters else {}
+
+    if user_filters:
+        user_filters = codec.encode_filter(user_filters)
+
+        def with_user_filters(encoded_filter: dict[str, Any]) -> dict[str, Any]:
+            return {"$and": [encoded_filter, encoded_user_filters]}
+    else:
+
+        def with_user_filters(encoded_filter: dict[str, Any]) -> dict[str, Any]:
+            return encoded_filter
+
+    parts = []
+    for k, v in metadata.items():
+        for v_batch in batched(v, 100):
+            batch = list(v_batch)
+            if len(batch) == 0:
+                continue
+            if len(batch) == 1:
+                parts.append({k: batch[0]})
+            else:
+                parts.append({k: {"$in": batch}})
+    if len(parts) == 1:
+        yield with_user_filters(codec.encode_filter(parts[0]))
+    elif len(parts) > 0:
+        yield with_user_filters(codec.encode_filter({"$or": parts}))
```

Comment:
Wouldn't this need to be made into a "one yeld per part"? I would say that right now it is still preparing a single query with the OR trick to bypass guardrails.

Reply:
Yes. I was trying to split it into steps. Although, this is not doing the `$or` on `_id` (that is now separate yields). This relates to the "is it 100 in a list that we should split" or "100 conditions accumulated", etc. I'll likely revisit that, but possibly in a follow-up.